Add error repro test for FSDP ignored modules with mixed precision

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89971
* #89930

The ignored modules are still using the original precision, which
leads to the following error.

```
RuntimeError: mat1 and mat2 must have the same dtype
```

This is not blocking me at the moment, but the fix seems not too
hard. We can add a pre-forward hook to each ignored module to
convert activations to original precision, and a post-forward hook
to convert it back to the specified precision.

Differential Revision: [D41661029](https://our.internmc.facebook.com/intern/diff/D41661029)

mrshenli committed:
Add error repro test for FSDP ignored modules with mixed precision

The ignored modules are still using the original precision, which
leads to the following error.

```
RuntimeError: mat1 and mat2 must have the same dtype
```

This is not blocking me at the moment, but the fix seems not too
hard. We can add a pre-forward hook to each ignored module to
convert activations to original precision, and a post-forward hook
to convert it back to the specified precision.

[ghstack-poisoned]

mrshenli committed:
Update on "Add error repro test for FSDP ignored modules with mixed precision"


The ignored modules are still using the original precision, which
leads to the following error.

```
RuntimeError: mat1 and mat2 must have the same dtype
```

This is not blocking me at the moment, but the fix seems not too
hard. We can add a pre-forward hook to each ignored module to
convert activations to original precision, and a post-forward hook
to convert it back to the specified precision.

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_mixed_precision.py:
```diff
@@ -756,5 +756,47 @@ def test_mixed_precision_e2e_full_shard(self):
+    @skip_if_lt_x_gpu(1)
+    def test_mixed_precision_with_ignored_module(self):
+        model = Model().cuda()
+        bfloat16 = MixedPrecision(param_dtype=torch.bfloat16)
```
As a heads up, you might want to use `float16` here if it also reproduces the error since otherwise you need to handle conditional BF16 support in CI:
https://github.com/pytorch/pytorch/blob/f99818fbc52b90851a3ef2af729901332ea11289/test/distributed/fsdp/test_fsdp_mixed_precision.py#L81-L83

awgu replied:
(Assuming that this test is meant to be landed)

mrshenli replied:
ah, good point, let me update.

mrshenli committed:
Update on "Add error repro test for FSDP ignored modules with mixed precision"


The ignored modules are still using the original precision, which
leads to the following error.

```
RuntimeError: mat1 and mat2 must have the same dtype
```

This is not blocking me at the moment, but the fix seems not too
hard. We can add a pre-forward hook to each ignored module to
convert activations to original precision, and a post-forward hook
to convert it back to the specified precision.

[ghstack-poisoned]

mrshenli commented:
@pytorchbot merge -g

mrshenli commented:
@pytorchbot merge -f "test failure is irrelevant"

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41661029).

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41661029).
